File: lcs/java/LCS.java
import java.util.*;

class LCS {
	private String s1;
	private String s2;

	public static void main (String[] args) {
		if (args.length != 2) {
			System.out.printf("Usage: LCS <string_1> <string_2>\n");
			System.exit(0);
		}

		LCS problem = new LCS (args[0], args[1]);;
		Map<MyIndex, Integer> allSubstrings = problem.findAllCommonSubstrings();
		List<String> lcSubStrings = problem.getLongestSubstrings(allSubstrings);
		System.out.printf("The longest common substrings are:\n");
		for (String substring : lcSubStrings) {
			System.out.printf("\t%s\n", substring);
		}
	}

	public LCS (String _s1, String _s2) {
		s1 = _s1;
		s2 = _s2;
	}

	private Map<MyIndex, Integer> findAllCommonSubstrings () {
		Map<MyIndex, Integer> collisions = new HashMap<>();

		int i = 0;
		for (Character c1 : s1.toCharArray()) {
			int j = 0;
			for (Character c2 : s2.toCharArray()) {
				if (c1 == c2) {
					MyIndex oldKey = new MyIndex(i-1, j-1);
					Integer previous = collisions.get(oldKey);
					if (previous == null) {previous = 0;}
					collisions.remove(oldKey);
					collisions.put(new MyIndex (i, j), previous + 1);
				}
			j++;
			}
		i++;
		}
		return collisions;
	}

	private List<String> getLongestSubstrings (Map<MyIndex, Integer> substrings) {
		List<String> longestSubstrings = new LinkedList<>();
		if (substrings == null || substrings.isEmpty()) {
			return longestSubstrings;
		}
		// get the maximum
		int maximum = 0;
		for (int value : substrings.values()) {
			if (value >= maximum) {maximum = value;}
		}

		// get the indices of the most common substrings from the hashmap
		// then look them up in the original string
		for (Map.Entry<MyIndex, Integer> entry : substrings.entrySet()) {
			if (entry.getValue().equals(maximum)) {

				int endIndex = entry.getKey().getFst() + 1;
				int startIndex = endIndex - entry.getValue();
				longestSubstrings.add(s1.substring(startIndex, endIndex));
			}
		}
		return longestSubstrings;
	}
}

class MyIndex {
	final private int fst, snd;

	public MyIndex (int _fst, int _snd) {
		fst = _fst;
		snd = _snd;
	}

	public int getFst() {return fst;}
	public int getSnd() {return snd;}

	@Override
	public boolean equals (Object o) {
		if (o == null) {return false;}
		if (!(o instanceof MyIndex)) {return false;}
		MyIndex pair = (MyIndex) o;
		return this.fst == pair.fst && this.snd == pair.snd;
	}

	@Override
	public int hashCode() {return fst*16777619 ^ snd;} // FNV hash
}

File: lcs/c/LCS.c

#define NDEBUG

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "uthash-master/src/uthash.h"
#include "uthash-master/src/utlist.h"

#include "LCS.h"

int main (int argc, char* argv[]) {
	if (argc != 3) {
		printf("Usage: %s <string_1> <string_2>", argv[0]);
		exit(0);
	}
	my_index** substrings = findAllCommonSubstrings (argv[1], argv[2]);
	substring_list* lcs = getLongestSubstrings (substrings, argv[1]);

	// free the hash table
	my_index *current_index, *tmp;
  	HASH_ITER(hh, *substrings, current_index, tmp) {
    	free(current_index);
    }
    HASH_CLEAR(hh, *substrings);

    // do stuff with the list of longest common substrings.
    // In this case we just print them.
	printf ("The longest common substrings are:\n");
	substring_list* tmp_lcs;
	LL_FOREACH (lcs, tmp_lcs) {
		printf("\t%s\n", tmp_lcs->substring);
	}


	// free the list
	substring_list* tmp_lcs_2;
	LL_FOREACH_SAFE(lcs, tmp_lcs, tmp_lcs_2) {
		LL_DELETE (lcs, tmp_lcs);
		free(tmp_lcs);
	}

}

my_index** findAllCommonSubstrings (const char* const restrict s1,
                                    const char* const restrict s2) {

	my_index** substrings = malloc(sizeof(my_index*));
	*substrings = NULL;

	int i;
	int iN = strlen(s1);
	int j;
	int jN = strlen(s2);

	for (i = 0; i < iN; i++) {
		for (j = 0; j < jN; j++) {
			if (s1[i] == s2[j]) {
				// find and delete the previous value, default of 0
				int old_key = hash (i-1, j-1);
				my_index* previous;
				HASH_FIND_INT (*substrings, &old_key, previous);
				int previous_value = 0;
				if (previous != NULL) {
					previous_value = previous->ss_length;
					HASH_DEL(*substrings, previous);
				}

				free(previous);

				// create the new value to be inserted
				int new_key = hash (i, j);
				my_index* new_value = malloc (sizeof(my_index));
				*new_value = (my_index){i, new_key, previous_value + 1};

				// if there's a collision, we have a problem
				my_index* new_index;
				HASH_FIND_INT (*substrings, &new_key, new_index);
				if (new_index != NULL) {
					printf ("Error: string2 is too long. 16777618 characters (~16MB) is the maximum");
				} else {
					HASH_ADD_INT(*substrings, key, new_value);
				}
			}
		}
	}
	return substrings;
}

substring_list* getLongestSubstrings (my_index** substrings,
                                   const char* const s) {
	substring_list* longest_substrings = NULL;

	// loop through the whole hash, looking for the maximum length
	int max_length = 0;
	my_index *current_index, *tmp;
  	HASH_ITER(hh, *substrings, current_index, tmp) {
  		int current_length = current_index->ss_length;
  		if (current_length > max_length) {max_length = current_length;}
    }

	// any strings with the max length get added to the list
	HASH_ITER(hh, *substrings, current_index, tmp) {
  		int current_length = current_index->ss_length;
  		if (current_length == max_length) {
  			// add the substring to the list
  			substring_list* new_ss = malloc (sizeof (substring_list));

  			size_t length = current_index->ss_length;
  			size_t startIndex = current_index->ix - length + 1;
  			new_ss->substring = malloc (length+1 * sizeof (char));
  			strncpy (new_ss->substring, s + startIndex, length);

  			LL_PREPEND(longest_substrings, new_ss);
  		}
    }

	// clear the whole hashtable
	return longest_substrings;
}

int hash (int _ix1, int _ix2) {
	return _ix1*16777619 ^ _ix2; // FNV hash
}